layers: Add shader output variables assignment check

scyganINTEL opened the pull request:
According to spec it is invalid to use input variable in consumer
stage, when no value was assigned to it in previous stages.

This change add related validation checks to core layer.

scygan committed:
layers: Add shader output variables assignment check

According to spec it is invalid to use input variable in consumer
stage, when no value was assigned to it in previous stages.

This change add related validation checks to core layer.

chrisforbes commented on layers/core_validation.cpp:
```diff
@@ -1662,6 +1662,102 @@ static std::vector<std::pair<descriptor_slot_t, interface_var>> collect_interfac
+                            for (auto it_interface = interface_variables.begin(); it_interface != interface_variables.end(); ++it_interface){
+                                if (it_interface->second.id == used_id){
+                                    if (!ids_usage.count(it_interface->first))
+                                        ids_usage[it_interface->first] = unordered_set<uint32_t>();
```
You don't need to do this explicitly-- operator[] invoked on the next line default-constructs a missing entry.

chrisforbes commented on layers/core_validation.cpp:
```diff
@@ -1662,6 +1662,102 @@ static std::vector<std::pair<descriptor_slot_t, interface_var>> collect_interfac
+    /* for every location loaded by consumer, check if it was stored by producer */
+    for (auto consumer_loaded_val : consumer_loaded_vals) {
+        if (!producer_saved_vals.count(consumer_loaded_val.first) && !producer_saved_pointers.count(consumer_loaded_val.first)) {
+            if (log_msg(report_data, VK_DEBUG_REPORT_PERFORMANCE_WARNING_BIT_EXT, VkDebugReportObjectTypeEXT(0), 0,
```
Is perf warning the right strength?

chrisforbes commented:
First wave of comments is just surface stuff from a really brief look. I need to think about this a bit more.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/Vulkan-LoaderAndValidationLayers?pullRequest=1108) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/Vulkan-LoaderAndValidationLayers?pullRequest=1108) before we can accept your contribution.<br/><hr/>**tmacyszx** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

scyganINTEL commented:
Improved performance by eliminating interface variables' copy, unnecessary iteration and copying of redundant arguments. Also changed the report flag.

Removed atomics used against pointers.

scygan commented:
@chrisforbes  - requested changes have been implemented. 

Can we have another round of review?

chrisforbes commented on layers/core_validation.cpp:
```diff
@@ -1662,6 +1662,104 @@ static std::vector<std::pair<descriptor_slot_t, interface_var>> collect_interfac
 }
 
+/* Returns locations mapped to offsets in entrypoint where they were used */
+map<location_t , unordered_set<uint32_t>> collect_interface_usage(shader_module const *src, map<location_t, interface_var>& interface_variables, 
```
`interface_variables` should be ref to const.

chrisforbes reviewed (commented):
This looks fairly sensible to me. It needs tests, though.

tobine reviewed (changes requested):
The code looks reasonable but it's currently only building for windows. Large streams of error messages for both Linux and Android that look to be centered around the use of spv::Op in unordered_map. Perhaps a hash and equality functions definitions are required?

mark-lunarg commented:
Labeled 'inactive' and closing -- please reopen if necessary.
